### PR TITLE
Separate function bodies from their signatures in HIR

### LINKED ISSUE #35078: Separate fn signature from fn body in HIR
Currently, getting access to a HIR item gives you access to both the function signature _and_ its body. This makes the incremental compilation graph more coarse than it should be. To fix this, the best thing would be to separate things out so that the item is just the external signature, and you have to separately load (e.g., from the HIR map) the body for a fn item. This way we could record that you actually accessed the fn body and track that separately.

### PR DESCRIPTION
**Pull request by flodiebold**
Also give them their own dep map node.

I'm still unhappy with the handling of inlined items (1452edc1), but maybe you have a suggestion how to improve it.

Fixes #35078.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_incremental/calculate_svh/mod.rs
Comment (nikomatsakis) [inline src/librustc_incremental/calculate_svh/mod.rs]: If we were using `DepNode::Hir` for everything, then this code would not (I think) have to change; we would probably just never walk bodies in the hasher. Instead, we would call `calculate_def_id` on the bodies of fns as we traversed.

File: src/test/incremental/hello_world.rs
Comment (nikomatsakis) [inline src/test/incremental/hello_world.rs]: Nice =)

File: src/test/incremental/source_loc_macros.rs
Comment (nikomatsakis) [inline src/test/incremental/source_loc_macros.rs]: Seems like it'd be good to have both `Hir` and `HirBody` here?

File: src/librustc/middle/cstore.rs
Comment (nikomatsakis) [inline src/librustc/middle/cstore.rs]: Is this a pre-existing bug, or something you introduced?

File: src/librustc_typeck/collect.rs
Comment (nikomatsakis) [inline src/librustc_typeck/collect.rs]: Hmm, this is suboptimal from the POV of the dep-graph tasks. We don't want the task which writes to the signature of a fn also reading from the fn body. Not sure what's the best way to tweak that just now; it's certainly possible.

File: src/test/incremental/hello_world.rs
Comment (nikomatsakis) [inline src/test/incremental/hello_world.rs]: I think the problem is the issue of collect reading from the fn body that I pointed out earlier.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
OK, this is looking really good! I had a few comments as to possible other ways that we could do things. I'm not sure which, if any, we should do before landing. I might have to experiment locally to form a real opinion in any case. Curious to get your opinion. These are the major themes I remember:

- [ ] use a separate method for determining whether to visit bodies or not?
- [ ] should we use `DepNode::HirBody`, or else give def-ids to fn bodies and use `DepNode::Hir`?
- [x] how to fix the collect edges? -- *pushed some commits into the branch that fix it*

I have some thoughts on the last one, but I think it'd be easier for me to just experiment locally and send you a diff if what I am saying makes sense. =)

File: src/librustc/middle/cstore.rs
Comment (eddyb) [inline src/librustc/middle/cstore.rs]: Seems okay - shouldn't need the type of a constant though, in fact, you should be able to only record a `Vec<DefId>` for the arguments of a `const fn`, which doesn't even need to be in a separate enum anymore, just replace the `kind` field with `const_fn_args: Vec<DefId>` and that should be that.

**Comment by flodiebold** [comment]
Giving def-ids to bodies didn't really occur to me (I was thinking only definitions can have def-ids). That would probably make a few things cleaner, but I don't really have the perspective to know what else it may complicate. Also, how would the testing attributes work? Can you give attributes to expressions?

File: src/librustc/middle/cstore.rs
Comment (flodiebold) [inline src/librustc/middle/cstore.rs]: The code which calls `from_trait_item` takes care to only call it when there's a body, but I accidentally changed that once, and it wasn't immediately clear to me what I'd done wrong, so I just added this special message in case it happens again.

File: src/librustc_passes/consts.rs
Comment (flodiebold) [inline src/librustc_passes/consts.rs]: Ah, yeah. Because the `FnLike` interface is only half implemented for inlined items, the `fn_like` call would fail for them, but the `const_qualif_map` is already populated for them, so I added the check beforehand. I'll need to clean that up somehow...

**Comment by nikomatsakis** [comment]
@flodiebold so I pushed two commits into your branch -- take a look. The first one fixes the `TypeckCollect` dependency problem in a reasonable clean way. The other one is a WIP -- basically a bunch of the "svh" tests are now much cleaner (in particular, a lot of once-dirty metadata and signatures are now clean) and I didn't get time to make them all pass.

**Comment by nikomatsakis** [comment]
> Can you give attributes to expressions?

You can give attributes to statements, not expressions. So that is a possible complication; I guess we could look for an attribute on the first statement of a block or something.

**Comment by nikomatsakis** [comment]
@flodiebold all in all, the main thing I'd prefer to settle now is the visitor trait -- i.e., whether to have another method. The `DepNode::HirBody` thing is relatively easy to refactor later, I think.

**Comment by flodiebold** [comment]
@nikomatsakis It does seem to me like having two methods and having no default for `nested_visit_map` would be cleaner. I'll do that.

I'll fix the remaining tests; I think they may have come in during my last rebase, after which I forgot to check them again :)

**Comment by nikomatsakis** [comment]
@flodiebold 

> It does seem to me like having two methods and having no default for nested_visit_map would be cleaner. I'll do that.

Great =)

If that change is made and tests are working, I think this PR is ready to go. We can revisit the `HirBody` thing afterwards. I tend to think it'd be mildly cleaner to just have one variant (`Hir`) but I think the diff will be non-trivial-enough we might as well handle it in a follow-up (we have to create def-ids for fn bodies, at minimum? I feel like when I thought it through I decided there would be some (minor, but still) complications to overcome.)

File: src/librustc/middle/cstore.rs
Comment (flodiebold) [inline src/librustc/middle/cstore.rs]: That would make implementing `FnLike` even harder, though, I think.

Comment (flodiebold) [inline src/librustc/middle/cstore.rs]: `lookup_const_fn_by_id` returns an `FnLikeNode` currently. But I guess it could return something else, e.g. a new enum?

Comment (eddyb) [inline src/librustc/middle/cstore.rs]: Yeah, that would make more sense, `FnLikeNode` is more about pretty-printing.

**Comment by flodiebold** [comment]
@eddyb I refactored the inlined item handling again. I didn't get rid of the `FnLikeNode` usage for local const fns though, just for the inlined items.

@nikomatsakis Ok, I think I've addressed everything.

**Comment by flodiebold** [comment]
@michaelwoerister By the way, the `unary_and_binary_exprs` test was actually not working correctly because `cfail2` was misspelled `cfails2` in the `rustc_clean` annotations. Maybe that could be prevented somehow?

**Comment by michaelwoerister** [comment]
@flodiebold Good catch. Yes, maybe we can let the rest runner do some checks to prevent this.

**Comment by eddyb** [comment]
@flodiebold The commit "Merge master into separate-bodies" is worrying me - did you `git pull` without `--rebase`? In other news, I'm blocked on this PR, more or less, so let me know if I can help.

**Comment by flodiebold** [comment]
@eddyb I did merge because I wasn't sure whether a merge or a rebase is the preferred way to resolve conflicts. But I've actually just done a rebase onto the current master, so I could push that too.

Apart from that, I think the PR is ready and just needs to be looked at again.

**Comment by eddyb** [comment]
@flodiebold Rebase + force push is the way, only @bors does merges (into the `auto` branch).
I've also used merge *locally* to temporarily create a branch made from multiple PRs to test something.

**Comment by flodiebold** [comment]
Ok, that's what I preferred anyway. Rebased version coming up.

**Comment by nikomatsakis** [comment]
@flodiebold I was experimenting with an alternative version of the visitor refactoring. I've just rebased it atop your rebase, once it builds I'll push it to your branch and you can tell me what you think. It's actually sort of closer to your original approach.

**Comment by nikomatsakis** [comment]
done.

**Comment by nikomatsakis** [comment]
Pushed one more commit updating the comments a bit.

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@flodiebold (if you are happy with my latest commit, then r=me on the PR)

**Comment by flodiebold** [comment]
Ah, that does seem nicer. Makes it easier to search for visitors which currently don't return a map, too. :+1:

**Comment by flodiebold** [comment]
@eddyb @nikomatsakis Fixed and rebased :)

**Comment by nikomatsakis** [comment]
@flodiebold great, thanks!

**Pull request merged**
